calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+                if !key.is_written(dominators, comp, &self.state_shareable) {
+                    self.no_share.insert(comp.name.clone());
+                    return Ok(Action::Continue);
+                }
+            }
+        }
+        comp.attributes.insert("state_share", 1);
+        self.state_shareable.add(comp.name.clone());
+        Ok(Action::Continue)
+    }
+}
+
+///Contains the ids of all the cells that are read from in a given "node" in
+///the domination map graph
+#[derive(Default)]
+struct ReadSet {
+    pub reads: HashSet<ir::Id>,
+}
+impl ReadSet {
+    //given a port, insert the port's parent's id if the parent the port's parent
+    //is shareable
+    fn add_parent_if_shareable(
+        &mut self,
+        share: &ShareSet,
+        port: &ir::RRC<ir::Port>,
+    ) {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
+            if share.is_shareable_component(&cell.upgrade()) {
+                self.reads.insert(cell.upgrade().borrow().name().clone());
+            }
+        }
+    }
+
+    //if the assignment reads only dones, return true. This is used so that we
+    //can ignore reads of "done" cells.
+    fn reads_only_dones(assignment: &ir::Assignment) -> bool {
+        Self::is_done_port(&assignment.src)
+            && Self::guard_only_dones(&assignment.guard)
+    }
+
+    //returns true if port is a "done" port or is a constant
+    fn is_done_port(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().name.id == "done" || port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard only contains dones, or is true
+    fn guard_only_dones(guard: &ir::Guard) -> bool {
+        match guard {
+            ir::Guard::Or(g1, g2) | ir::Guard::And(g1, g2) => {
+                Self::guard_only_dones(g1) && Self::guard_only_dones(g2)
+            }
+            ir::Guard::Not(g) => Self::guard_only_dones(g),
+            ir::Guard::True => true,
+            ir::Guard::CompOp(_, p1, p2) => {
+                Self::is_done_port(p1) && Self::is_done_port(p2)
+            }
+            ir::Guard::Port(p) => Self::is_done_port(p),
+        }
+    }
+
+    //Adds the ids of any state_shareable cells that are read from in assignments,
+    //excluding reads where the only reads are from "done" ports.
+    fn add_assignment_reads(
+        &mut self,
+        share: &ShareSet,
+        assignments: &[ir::Assignment],
+    ) {
+        for cell in ReadWriteSet::read_set(
+            assignments
+                .iter()
+                .filter(|assign| !Self::reads_only_dones(assign)),
+        ) {
+            if share.is_shareable_component(&cell) {
+                self.reads.insert(cell.borrow().name().clone());
+            }
+        }
+    }
+
+    //Given a control statement c, adds all of the reads of shareable cells from c.
+    //For while loops and if stmts, the control refers only to the guard, not the body.
+    fn get_reads_from_control(
```

Comment:
The `ReadWriteSet::control_port_read_write_set` defines a way to get all the reads and writes from a control tree

Reply:
I've taken a shot at fixing it. See if it still works